python/tests/io/test_pyarrow.py

```diff
@@ -1345,3 +1374,655 @@ def test_pyarrow_wrap_fsspec(example_task: FileScanTask, table_schema_simple: Sc
+                    {
+                        "id": 3,
+                        "name": "list",
+                        "required": False,
+                        "type": {"type": "list", "element-id": 5, "element": "long", "element-required": False},
+                    },
+                    {
+                        "id": 4,
+                        "name": "maps",
+                        "required": False,
+                        "type": {
+                            "type": "map",
+                            "key-id": 6,
+                            "key": "long",
+                            "value-id": 7,
+                            "value": "long",
+                            "value-required": False,
+                        },
+                    },
+                ],
+            },
+        ],
+        "default-spec-id": 0,
+        "partition-specs": [{"spec-id": 0, "fields": []}],
+        "properties": {},
+    }
+
+    table_metadata = TableMetadataUtil.parse_obj(table_metadata)
+    arrow_schema = schema_to_pyarrow(table_metadata.schemas[0])
+
+    _strings = ["zzzzzzzzzzzzzzzzzzzz", "rrrrrrrrrrrrrrrrrrrr", None, "aaaaaaaaaaaaaaaaaaaa"]
+
+    _floats = [3.14, math.nan, 1.69, 100]
+
+    _list = [[1, 2, 3], [4, 5, 6], None, [7, 8, 9]]
+
+    _maps: List[Optional[Dict[int, int]]] = [
+        {1: 2, 3: 4},
+        None,
+        {5: 6},
+        {},
+    ]
+
+    table = pa.Table.from_pydict(
+        {
+            "strings": _strings,
+            "floats": _floats,
+            "list": _list,
+            "maps": _maps,
+        },
+        schema=arrow_schema,
+    )
+    metadata_collector: List[Any] = []
+
+    with pa.BufferOutputStream() as f:
+        with pq.ParquetWriter(f, table.schema, metadata_collector=metadata_collector) as writer:
+            writer.write_table(table)
+
+        return f.getvalue(), metadata_collector[0], table_metadata
+
+
+def test_record_count() -> None:
+    (file_bytes, metadata, table_metadata) = construct_test_table()
+
+    datafile = DataFile()
+    fill_parquet_file_metadata(datafile, metadata, len(file_bytes), table_metadata)
+
+    assert datafile.record_count == 4
+
+
+def test_file_size() -> None:
+    (file_bytes, metadata, table_metadata) = construct_test_table()
+
+    datafile = DataFile()
+    fill_parquet_file_metadata(datafile, metadata, len(file_bytes), table_metadata)
+
+    assert datafile.file_size_in_bytes == len(file_bytes)
+
+
+def test_value_counts() -> None:
+    (file_bytes, metadata, table_metadata) = construct_test_table()
+
+    datafile = DataFile()
+    fill_parquet_file_metadata(datafile, metadata, len(file_bytes), table_metadata)
+
+    assert len(datafile.value_counts) == 5
+    assert datafile.value_counts[1] == 4
+    assert datafile.value_counts[2] == 4
+    assert datafile.value_counts[5] == 10  # 3 lists with 3 items and a None value
+    assert datafile.value_counts[6] == 5
```

Comment:
It looks like both maps and lists count empty/null parent structures. I would expect this to be the number of leaf values.

Can you confirm that this is also the behavior of the Java Parquet library? If not, we should just drop map and list stats because they aren't the same.

Reply:
With Spark:

```sql
CREATE TABLE nyc.test_map_maps2 AS 
SELECT map_from_arrays(array(1.0, 3.0), array('2', '4')) as map, array('a', 'b', 'c') as arr
```

Schema:
```json
{
	"type": "struct",
	"schema-id": 0,
	"fields": [{
		"id": 1,
		"name": "map",
		"required": false,
		"type": {
			"type": "map",
			"key-id": 3,
			"key": "decimal(2, 1)",
			"value-id": 4,
			"value": "string",
			"value-required": false
		}
	}, {
		"id": 2,
		"name": "arr",
		"required": false,
		"type": {
			"type": "list",
			"element-id": 5,
			"element": "string",
			"element-required": false
		}
	}]
}
```

We don't get any stats in Spark:
```json
{
	"status": 1,
	"snapshot_id": {
		"long": 4895801649705337905
	},
	"data_file": {
		"file_path": "s3://warehouse/nyc/test_map_maps/data/00000-95-750d8f3e-8d49-44ec-b37e-9e101e003a5d-00001.parquet",
		"file_format": "PARQUET",
		"partition": {},
		"record_count": 1,
		"file_size_in_bytes": 1438,
		"block_size_in_bytes": 67108864,
		"column_sizes": {
			"array": [{
				"key": 3,
				"value": 57
			}, {
				"key": 4,
				"value": 58
			}, {
				"key": 5,
				"value": 61
			}]
		},
		"value_counts": {
			"array": [{
				"key": 3,
				"value": 2
			}, {
				"key": 4,
				"value": 2
			}, {
				"key": 5,
				"value": 3
			}]
		},
		"null_value_counts": {
			"array": [{
				"key": 3,
				"value": 0
			}, {
				"key": 4,
				"value": 0
			}, {
				"key": 5,
				"value": 0
			}]
		},
		"nan_value_counts": {
			"array": []
		},
		"lower_bounds": {
			"array": []
		},
		"upper_bounds": {
			"array": []
		},
		"key_metadata": null,
		"split_offsets": {
			"array": [4]
		},
		"sort_order_id": {
			"int": 0
		}
	}
}
```

The stats are only computed for the primitive types (see `PyArrowStatisticsCollector`).

Reply:
```sql
CREATE TABLE default.arrays
SELECT
    array(1, 2, 3, null) as with_a_null,
    array(1, 2, cast('NaN' as double), null) as with_null_and_nan,
    array(1, 2, 3, cast('NaN' as double)) as with_a_nan
```

Schema:
```json
{
	"type": "struct",
	"schema-id": 0,
	"fields": [{
		"id": 1,
		"name": "with_a_null",
		"required": false,
		"type": {
			"type": "list",
			"element-id": 4,
			"element": "int",
			"element-required": false
		}
	}, {
		"id": 2,
		"name": "with_null_and_nan",
		"required": false,
		"type": {
			"type": "list",
			"element-id": 5,
			"element": "double",
			"element-required": false
		}
	}, {
		"id": 3,
		"name": "with_a_nan",
		"required": false,
		"type": {
			"type": "list",
			"element-id": 6,
			"element": "double",
			"element-required": false
		}
	}]
}
```

The missing upper- and lower bound seems to be an issue with Spark. It does generates the doubles:

```json
{
	"status": 1,
	"snapshot_id": {
		"long": 3911973389432551915
	},
	"data_file": {
		"file_path": "s3://warehouse/default/arrays/data/00000-0-1e0d4a20-4473-450e-9b1e-9196b5edac66-00001.parquet",
		"file_format": "PARQUET",
		"partition": {},
		"record_count": 1,
		"file_size_in_bytes": 1310,
		"block_size_in_bytes": 67108864,
		"column_sizes": {
			"array": [{
				"key": 4,
				"value": 60
			}, {
				"key": 5,
				"value": 63
			}, {
				"key": 6,
				"value": 66
			}]
		},
		"value_counts": {
			"array": [{
				"key": 4,
				"value": 4
			}, {
				"key": 5,
				"value": 4
			}, {
				"key": 6,
				"value": 4
			}]
		},
		"null_value_counts": {
			"array": [{
				"key": 4,
				"value": 1
			}, {
				"key": 5,
				"value": 1
			}, {
				"key": 6,
				"value": 0
			}]
		},
		"nan_value_counts": {
			"array": [{
				"key": 5,
				"value": 1
			}, {
				"key": 6,
				"value": 1
			}]
		},
		"lower_bounds": {
			"array": [{
				"key": 5,
				"value": "\u0000\u0000\u0000\u0000\u0000\u0000ð?"
			}, {
				"key": 6,
				"value": "\u0000\u0000\u0000\u0000\u0000\u0000ð?"
			}]
		},
		"upper_bounds": {
			"array": [{
				"key": 5,
				"value": "\u0000\u0000\u0000\u0000\u0000\u0000\u0000@"
			}, {
				"key": 6,
				"value": "\u0000\u0000\u0000\u0000\u0000\u0000\b@"
			}]
		},
		"key_metadata": null,
		"split_offsets": {
			"array": [4]
		},
		"sort_order_id": {
			"int": 0
		}
	}
}
```

Reply:
Spark has the same number of values:
![image](https://github.com/apache/iceberg/assets/1134248/ad3a7f71-2e7d-46af-8309-c2720b97ff75)

Reply:
Seems to be consistent with what we're seeing in python